zcash_client_backend/src/proto.rs

```diff
@@ -183,3 +204,223 @@ impl service::TreeState {
+    {
+        match self.proto_version {
+            PROPOSAL_SER_V1 => {
+                #[allow(deprecated)]
+                let fee_rule = match self.fee_rule() {
+                    proposal::FeeRule::PreZip313 => StandardFeeRule::PreZip313,
+                    proposal::FeeRule::Zip313 => StandardFeeRule::Zip313,
+                    proposal::FeeRule::Zip317 => StandardFeeRule::Zip317,
+                    proposal::FeeRule::NotSpecified => {
+                        return Err(ProposalError::FeeRuleNotSpecified);
+                    }
+                };
+
+                let transaction_request =
+                    TransactionRequest::from_uri(params, &self.transaction_request)?;
+                let transparent_inputs = self
+                    .transparent_inputs
+                    .iter()
+                    .map(|t_in| {
+                        let txid = t_in.parse_txid().map_err(ProposalError::TxIdInvalid)?;
+                        let outpoint = OutPoint::new(txid.into(), t_in.index);
+
+                        wallet_db
+                            .get_unspent_transparent_output(&outpoint)
+                            .map_err(ProposalError::InputRetrieval)?
+                            .ok_or_else(|| {
+                                ProposalError::InputNotFound(
+                                    txid,
+                                    PoolType::Transparent,
+                                    t_in.index,
+                                )
+                            })
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                let sapling_inputs = self.sapling_inputs.as_ref().and_then(|s_in| {
+                    s_in.inputs
+                        .iter()
+                        .map(|s_in| {
+                            let txid = s_in.parse_txid().map_err(ProposalError::TxIdInvalid)?;
+
+                            wallet_db
+                                .get_spendable_sapling_note(&txid, s_in.index)
+                                .map_err(ProposalError::InputRetrieval)
+                                .and_then(|opt| {
+                                    opt.ok_or_else(|| {
+                                        ProposalError::InputNotFound(
+                                            txid,
+                                            PoolType::Shielded(ShieldedProtocol::Sapling),
+                                            s_in.index,
+                                        )
+                                    })
+                                })
+                        })
+                        .collect::<Result<Vec<_>, _>>()
+                        .map(|notes| {
+                            NonEmpty::from_vec(notes).map(|notes| {
+                                SaplingInputs::from_parts(s_in.anchor_height.into(), notes)
+                            })
+                        })
+                        .transpose()
```

Comment:
This map is a bit weird. We already map over `self.sapling_inputs` which represents the optionality of all Sapling inputs. This `map` introduces a non-canonical encoding that can contain an anchor height but no notes, and the parser will drop the anchor height (meaning that this non-canonical encoding is not round-trip compatible). I feel like this case should instead be rejected with something like `ProposalError::NonCanonical`. I know that protobuf itself does not have canonical encodings (mainly due to unconstrained field ordering and backwards compatibility support), but I feel that in this particular case it is a structurally invalid representation (so maybe we don't call this "non-canonical" and instead some other kind of error).

Non-blocking thought prompted by this: we should set up proptests and/or a fuzzer for this proposal format.

Reply:
Yeah, this seems okay; we can treat a `SaplingInputs` that has an anchor height but no inputs as an error.